File: backend/workflow/bash.go
package workflow

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"os"
	"os/exec"
	"path/filepath"

	"github.com/enough/enough/backend/shell"
)

const maxWorkflowBashOutput = 64 * 1024

func (r *Runtime) runBash(ctx context.Context, command string) (BashResult, error) {
	if command == "" {
		return BashResult{}, errors.New("runBash command is empty")
	}
	cmd, err := shell.CommandContext(ctx, command, true)
	if err != nil {
		return BashResult{}, err
	}
	cmd.Dir = shell.ResolveSafeCwd(r.workDir)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	err = cmd.Run()
	exitCode := 0
	if err != nil {
		exitCode = -1
		var ee *exec.ExitError
		if errors.As(err, &ee) {
			exitCode = ee.ExitCode()
		} else if ctx.Err() != nil {
			return BashResult{}, ctx.Err()
		}
	}
	result := BashResult{Stdout: stdout.String(), Stderr: stderr.String(), ExitCode: exitCode}
	combined := append(append([]byte(nil), stdout.Bytes()...), stderr.Bytes()...)
	if len(combined) > maxWorkflowBashOutput {
		hash := sha256.Sum256(combined)
		result.SHA256 = hex.EncodeToString(hash[:])
		result.Truncated = true
		dir := filepath.Join(filepath.Dir(r.snapshot.ScriptPath), "artifacts")
		_ = os.MkdirAll(dir, 0o755)
		path := filepath.Join(dir, "bash-"+result.SHA256[:16]+".log")
		_ = os.WriteFile(path, combined, 0o600)
		result.FullOutputPath = path
		if len(result.Stdout) > maxWorkflowBashOutput/2 {
			result.Stdout = result.Stdout[:maxWorkflowBashOutput/2] + "\n... truncated; full output: " + path
		}
		if len(result.Stderr) > maxWorkflowBashOutput/2 {
			result.Stderr = result.Stderr[:maxWorkflowBashOutput/2] + "\n... truncated; full output: " + path
		}
	}
	return result, nil
}

File: backend/workflow/bash.ts
// PORT: backend/workflow/bash.go

import { Effect } from "effect";
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { Runtime } from "./runtime";
import { type BashResult } from "./types";
import { command_context } from "../shell/command";
import { resolve_safe_cwd } from "../shell/cwd";
import { killProcessTree } from "../shell/process_tree";

const maxWorkflowBashOutput = 64 * 1024;

export function runBash(
  this: Runtime,
  ctx: AbortSignal,
  command: string
): Effect.Effect<BashResult, Error> {
  if (command === "") {
    return Effect.fail(new Error("runBash command is empty"));
  }

  const safeDir = resolve_safe_cwd(this.workDir);
  const commandEff = command_context(ctx, command, true, safeDir).pipe(
    Effect.mapError((err) => (err instanceof Error ? err : new Error(String(err))))
  );

  return commandEff.pipe(
    Effect.flatMap((child) =>
      Effect.async<BashResult, Error>((resume) => {
        const stdoutChunks: Buffer[] = [];
        const stderrChunks: Buffer[] = [];

        child.stdout?.on("data", (chunk: Buffer) => {
          stdoutChunks.push(chunk);
        });
        child.stderr?.on("data", (chunk: Buffer) => {
          stderrChunks.push(chunk);
        });

        let completed = false;
        let onAbort: () => void;

        const cleanUp = () => {
          if (completed) return false;
          completed = true;
          ctx.removeEventListener("abort", onAbort);
          return true;
        };

        const finish = (code: number | null, signal: string | null) => {
          if (!cleanUp()) return;

          const stdoutBuf = Buffer.concat(stdoutChunks);
          const stderrBuf = Buffer.concat(stderrChunks);
          const stdoutStr = stdoutBuf.toString("utf8");
          const stderrStr = stderrBuf.toString("utf8");
          const exitCode = code !== null ? code : -1;

          const combined = Buffer.concat([stdoutBuf, stderrBuf]);
          const result: BashResult = {
            stdout: stdoutStr,
            stderr: stderrStr,
            exitCode: exitCode,
          };

          if (combined.length > maxWorkflowBashOutput) {
            const hash = crypto.createHash("sha256").update(combined).digest("hex");
            result.sha256 = hash;
            result.truncated = true;

            const dir = path.join(path.dirname(this.snapshot.scriptPath), "artifacts");
            try {
              fs.mkdirSync(dir, { recursive: true, mode: 0o755 });
            } catch {}

            const logPath = path.join(dir, `bash-${hash.slice(0, 16)}.log`);
            try {
              fs.writeFileSync(logPath, combined, { mode: 0o600 });
            } catch {}

            result.fullOutputPath = logPath;

            const halfMax = Math.floor(maxWorkflowBashOutput / 2);
            if (stdoutStr.length > halfMax) {
              result.stdout = stdoutStr.slice(0, halfMax) + "\n... truncated; full output: " + logPath;
            }
            if (stderrStr.length > halfMax) {
              result.stderr = stderrStr.slice(0, halfMax) + "\n... truncated; full output: " + logPath;
            }
          }

          resume(Effect.succeed(result));
        };

        child.on("exit", (code, signal) => {
          finish(code, signal);
        });

        child.on("close", (code, signal) => {
          finish(code, signal);
        });

        child.on("error", (err) => {
          if (!cleanUp()) return;
          resume(Effect.fail(err));
        });

        onAbort = () => {
          if (!cleanUp()) return;
          if (child.pid) killProcessTree(child.pid);
          resume(Effect.fail(new Error("interrupted")));
        };

        if (ctx.aborted) {
          onAbort();
        } else {
          ctx.addEventListener("abort", onAbort);
        }
      })
    )
  );
}

(Runtime.prototype as any).runBash = runBash;

/*
PORT STATUS
source path: backend/workflow/bash.go
source lines: 61
draft lines: 111
confidence: high
status: phase_b_compile
*/
